core/trie/proof.go

```diff
@@ -133,6 +133,163 @@ func transformNode(tri *Trie, parentKey *Key, sNode StorageNode) (*Edge, *Binary
+			expectedRightHash := currNode.Binary.RightHash.String()
+
+			nodeLeft, leftExist := nodeHashes[expectedLeftHash]
+			nodeRight, rightExist := nodeHashes[expectedRightHash]
+			if leftExist && rightExist {
+				return errors.New("unexpected error")
+			} else if leftExist {
+				currNode = &nodeLeft
+				*path = append(*path, *currNode)
+			} else if rightExist {
+				currNode = &nodeRight
+				*path = append(*path, *currNode)
+			} else {
+				whileLoop = false
+			}
+		case currNode.Edge != nil:
+			righNode, exists := nodeHashes[currNode.Edge.Child.String()]
+			if exists {
+				currNode = &righNode
+				*path = append(*path, *currNode)
+			} else {
+				whileLoop = false
+			}
+		}
+	}
+
+	return nil
+}
+
+// Remove duplicates and merges proof paths into a single path
+func MergeProofPaths(leftPath, rightPath []ProofNode, hash hashFunc) ([]ProofNode, *felt.Felt, error) {
+	merged := []ProofNode{}
+	minLen := min(len(leftPath), len(rightPath))
+
+	if len(leftPath) == 0 || len(rightPath) == 0 {
+		return merged, nil, errors.New("empty proof paths")
+	}
+
+	if !leftPath[0].Hash(hash).Equal(rightPath[0].Hash(hash)) {
+		return merged, nil, errors.New("roots of the proof paths are different")
+	}
+
+	rootHash := leftPath[0].Hash(hash)
+
+	// Get duplicates and insert by one
+	i := 0
+	for i = 0; i < minLen; i++ {
+		leftNode := leftPath[i]
+		rightNode := rightPath[i]
+
+		if leftNode.Hash(hash).Equal(rightNode.Hash(hash)) {
+			merged = append(merged, leftNode)
+		} else {
+			break
+		}
+	}
+
+	// Add rest of the nodes one from left and one from right
+	// until we reach the end of the shortest path
+	for ; i < minLen; i++ {
+		merged = append(merged, leftPath[i], rightPath[i])
+	}
+
+	// Add the rest of the nodes from the longest path
+	if len(leftPath) > minLen {
+		merged = append(merged, leftPath[i:]...)
+	} else if len(rightPath) > minLen {
+		merged = append(merged, rightPath[i:]...)
+	}
+
+	return merged, rootHash, nil
+}
+
+func SplitProofPath(mergedPath []ProofNode, rootHash *felt.Felt, hash hashFunc) ([]ProofNode, []ProofNode, error) {
+	leftPath := []ProofNode{}
+	rightPath := []ProofNode{}
+
+	if len(mergedPath) == 0 {
+		return leftPath, rightPath, nil
+	}
+
+	nodeHashes := make(map[string]ProofNode)
+
+	for _, node := range mergedPath {
+		nodeHash := node.Hash(hash)
+		nodeHashes[nodeHash.String()] = node
+	}
+
+	currNode, rootExists := nodeHashes[rootHash.String()]
+	if !rootExists {
+		return leftPath, rightPath, errors.New("root hash not found in the merged path")
+	}
+
+	leftPath = append(leftPath, currNode)
+	rightPath = append(rightPath, currNode)
+
+	whileLoop := true
+	for whileLoop {
+		switch {
+		case currNode.Binary != nil:
+			expectedLeftHash := currNode.Binary.LeftHash
+			expectedRightHash := currNode.Binary.RightHash
+
+			nodeLeft, leftExist := nodeHashes[expectedLeftHash.String()]
+			nodeRight, rightExist := nodeHashes[expectedRightHash.String()]
+			if leftExist && rightExist {
```

Comment:
Both left & right exist so this is the last "common" node.
What we do next is we're giving **this node** to `traverseNodesAndAppendToPath` (line 281 & 285 👇 )
and then in traverse it should blow up on this:

https://github.com/NethermindEth/juno/blob/e4c49a46cea749649969fb9999cb8688c4272fef/core/trie/proof.go#L146-L147

So this tells me we have no test with common path 🤔

Reply:
Yes we don't have a test with common path I will add it but also In common paths I guess both left and right paths will not exist in the same time since there will be no split

Reply:
This case actually shouldn't happen in traverse.

Reply:
Yes - my bad, split appends not the binary node that has both of left&right but the left & right themselves - so traverse works on them